### ISSUE TITLE
Make chaste_codegen version appear in autogenerated Version.cpp ?

### ISSUE THREAD

**Issue by mirams** [Chaste/Chaste#122]
Whereas other executables write their versions into the build-time-generated `Version.cpp`, chaste_codegen doesn't and does some weird looking file opening to read it. Think about whether we can put in the right info at build time.
```
const char* ChasteBuildInfo::GetCompilerVersion()
{
    return "9.4.0";
}

std::string ChasteBuildInfo::GetChasteCodegenVersion(){
    FileFinder chaste_build_root("", RelativeTo::ChasteBuildRoot);
    char buffer[100];
    std::string codgen_version_command = chaste_build_root.GetAbsolutePath() + "/codegen_python3_venv/bin/chaste_codegen --version";
    FILE* codegen_version = popen(codgen_version_command.c_str(), "r");
    std::string results = "";

    while (fgets(buffer, 100, codegen_version) != nullptr)
    {
        results +=  buffer;
    }
    pclose(codegen_version);

    boost::algorithm::trim(results);
    return results;
}

const char* ChasteBuildInfo::GetXsdVersion()
{
    return "4.0.0";
}
```

Now because chaste_codegen is used in dynamic loading at runtime there could be a good reason for the above? But it does seem to cause a bottleneck at times so we should see if it can be optimised if it needs to be runtime.

**Comment by mirams** [comment]
@jmpf @jonc125 @MauriceHendrix thoughts welcome. This is to do with issue #119

**Comment by MauriceHendrix** [comment]
> Whereas other executables write their versions into the build-time-generated `Version.cpp`, chaste_codegen doesn't and does some weird looking file opening to read it. Think about whether we can put in the right info at build time.
> 
I think you're going to struggle to put the run-time codegen version number in the C++, since it's a python package that you install seperately (although it's done automatically during the build process). It's perfectly possible to switch to a different version after the build. 

If you don't mind the chance of that descrepancy, then it's probably possible.

**Comment by fcooper8472** [comment]
I think you could technically switch other things, too. E.g. the version of XSD is hardcoded too:

```cpp
const char* ChasteBuildInfo::GetXsdVersion()
{
    return "${xsd_version}";
}
```
which is a fixed string after the CMake configure step. But it's exceedingly unlikely that anyone would replace the XSD executable between configuring and running Chaste.

Is there a realistic chance that someone would go into the virtual environment that's created at configure-time and change/update codegen without re-configuring?

My hunch is it ought to be safe to get CMake to fill in that string.

(As a side-note, now we've officially got rid of SCons support, I think we can get rid of `Version.cpp.in` and just keep `Version_cmake.cpp.in`.)

**Comment by MauriceHendrix** [comment]
Probably not in the end to be fair.

**Comment by kwabenantim** [comment]
Those were my thoughts as well @MauriceHendrix. It will probably only be developers changing the codegen version after Chaste has already been built so not a bad idea to put the version Chaste was configured with in `Version.cpp`.

**Comment by fcooper8472** [comment]
Closed by #123
